DBW: Audit for same-origin requests over http/2

ebidel opened the pull request:
R: @brendankenny @paulirish @GoogleChrome/lighthouse 

<img width="835" alt="screen shot 2016-09-23 at 10 19 37 am" src="https://cloud.githubusercontent.com/assets/238208/18795028/4c829c68-8177-11e6-9c95-c6e6d458585c.png">

ebidel commented:
@brendankenny for closure failures

brendankenny commented on lighthouse-core/audits/dobetterweb/uses-http2.js:
```diff
@@ -0,0 +1,75 @@
+
+/**
+ * @fileoverview Audit a page to ensure that resource loaded over its own
+ * origin are over the http/2 protocol.
```
why just same origin? Do you want to give that caveat somewhere in the helptext?

ebidel replied:
I figure there will be a lot of sites that are using h2 but make http1.1 requests to API endpoints, ad networks, etc. In those cases, there's not much the site author can do.

ebidel replied:
Added to helpText

brendankenny commented on lighthouse-core/audits/dobetterweb/uses-http2.js:
```diff
@@ -0,0 +1,75 @@
+
+    // Filter the non http/2 resources.
+    const resources = artifacts.SameOriginResources.filter(record => {
+      return /HTTP\/[01][\.\d]?/i.test(record.protocol);
```
any reason not to test this against `'h2'`?

ebidel replied:
Don't think so. "HTTP/2.0" is the compliant header.

brendankenny replied:
I don't think this field is the header, but the string identifier internal to devtools (which also looks like what's shown in the network panel). From a quick look at a few sites it's giving me `'http/1.1'` for requests from e.g. example.com and `'h2'` for requests from e.g. chromestatus

ebidel replied:
Right, and I only want the http/1.x results.

brendankenny commented on lighthouse-core/formatters/url-list.js:
```diff
@@ -0,0 +1,46 @@
+          if (!Array.isArray(resources)) {
+            return '';
+          }
+          // TODO
```
even something basic here would be preferable to TODO

ebidel replied:
Done

brendankenny commented on lighthouse-core/test/audits/dobetterweb/uses-http2-test.js:
```diff
@@ -0,0 +1,51 @@
+  it('passes when all resources were requested via http/2', () => {
+    const auditResult = UsesHTTP2Audit.audit({
+      SameOriginResources: [
+        {url: 'http://example.com/one', protocol: 'http/2'},
```
I believe our network records use `'h2'` for the protocol string

ebidel replied:
Done

brendankenny commented on lighthouse-core/test/audits/dobetterweb/uses-http2-test.js:
```diff
@@ -0,0 +1,51 @@
+
+/* eslint-env mocha */
+
+describe('Resources are fetched over http/2', () => {
```
since you're computing directly against the network records, it would be nice to test on an actual copy of the network records. There's one using h2 push in core/test/fixtures/networkRecords-h2push.json. I thought there was one more, but you can generate one via Lighthouse if you pass in the `--save-artifacts` flag when running against some site.

ebidel commented on lighthouse-core/gather/gatherers/same-origin-resources.js:
```diff
@@ -0,0 +1,44 @@
+const url = require('url');
+const Gatherer = require('./gatherer');
+
+class SameOriginResources extends Gatherer {
```
I wanted to just created an audit but I didn't see a way to get options for `options.initialUrl` and `options.url`. Do audits have that info?

brendankenny replied:
if you include the `url` gatherer it passes along the final url (`options.url`) on `artifacts.URL`

brendankenny replied:
the `initialUrl` isn't passed in, but a gatherer could do that pretty easily (or URL could do both or something)

ebidel commented on lighthouse-core/gather/gatherers/same-origin-resources.js:
```diff
@@ -0,0 +1,44 @@
+    const initialHost = url.parse(options.initialUrl).host;
+
+    // Find requests that are on the same origin as the page.
+    const results = tracingData.networkRecords.reduce((prev, record) => {
```
Needed a map+filter before but I moved more to the audit. Done.

ebidel commented:
PTAL. Finished the pretty formatter but didn't move the tests to using the actual network results. Not sure that's really needed to test the audit output.

brendankenny commented:
>  but didn't move the tests to using the actual network results. Not sure that's really needed to test the audit output

testing against a real network log is wayyy better than against data you wrote specifically for the test :P Plus it's super easy: you can `require()` the file and it's already the array of records you expect. See jeffposnick's [test for offline](https://github.com/GoogleChrome/lighthouse/blob/3350df49a858a8df153b389233e3a8161f46d6b8/lighthouse-core/test/gather/gatherers/offline-test.js)

brendankenny commented:
and there's another trace (at core/test/fixtures/traces/network-records.json). Unfortunately it also looks like it's h2-only

ebidel commented:
PTAL. Added http 1.1 mock data, removed the same origin gatherer (logic moved into the `uses-http2.js` audit, and refactored `gatherers/url.js` to pass the initial and final URL.

paulirish commented on lighthouse-core/audits/dobetterweb/uses-http2.js:
```diff
@@ -0,0 +1,83 @@
+    const resources = networkRecords.filter(record => {
+      const requestHost = url.parse(record._url).host;
+      const sameOrigin = requestHost === finalHost;
+      const notH2 = /HTTP\/[01][\.\d]?/i.test(record.protocol);
```
Does this handle quic requests? For example, I see `record.protocol === 'quic/1+spdy/3'` a lot.

ebidel replied:
I figured ignoring everything but http/[01].x would be ok. For all intents and purposes, quic/spdy is like h2 (more modern protocol and 👍  if in use).

paulirish commented on lighthouse-core/gather/gatherers/url.js:
```diff
@@ -25,7 +25,10 @@ class URL extends Gatherer {
     // Instead of the originally inputted URL (options.initialUrl), we want the resolved
     // post-redirect URL (which is here at options.url)
-    this.artifact = options.url;
+    this.artifact = {
```
This is a nice improvement. Thanks!

ebidel commented:
Can we land this? I have another PR that's blocked on these changes.

brendankenny commented on lighthouse-core/audits/dobetterweb/uses-http2.js:
```diff
@@ -0,0 +1,83 @@
+    // Filter requests that are on the same origin as the page and not over h2.
+    const resources = networkRecords.filter(record => {
+      const requestHost = url.parse(record._url).host;
+      const sameOrigin = requestHost === finalHost;
```
`sameHost`

ebidel replied:
done

brendankenny commented on lighthouse-core/test/gather/gatherers/service-worker-test.js:
```diff
@@ -40,7 +40,7 @@ describe('Service Worker gatherer', () => {
         }
       },
-      url
+      url: {finalUrl: url}
```
I believe `url` was correct here (this is the `options` object, not `artifacts`). No error was caused by changing this because the gatherer doesn't actually use `options.url`

ebidel replied:
done

ebidel commented on lighthouse-core/audits/dobetterweb/uses-http2.js:
```diff
@@ -0,0 +1,83 @@
+    const networkRecords = artifacts.networkRecords[Audit.DEFAULT_PASS];
+    const finalHost = url.parse(artifacts.URL.finalUrl).host;
+
+    // Filter requests that are on the same origin as the page and not over h2.
```
done

ebidel commented on lighthouse-core/audits/dobetterweb/uses-http2.js:
```diff
@@ -0,0 +1,83 @@
+/**
+ * @fileoverview Audit a page to ensure that resource loaded over its own
+ * origin are over the http/2 protocol.
+ * @author Eric Bidelman
```
done

ebidel commented:
👀 @brendankenny
